staging/src/k8s.io/client-go/tools/cache/shared_informer.go

```diff
@@ -559,7 +574,7 @@ func (s *sharedIndexInformer) AddEventHandlerWithResyncPeriod(handler ResourceEv
 		}
 	}
 
-	listener := newProcessListener(handler, resyncPeriod, determineResyncPeriod(resyncPeriod, s.resyncCheckPeriod), s.clock.Now(), initialBufferSize)
+	listener := newProcessListener(handler, resyncPeriod, determineResyncPeriod(resyncPeriod, s.resyncCheckPeriod), s.clock.Now(), initialBufferSize, s.HasSynced)
```

Comment:
hi @lavalamp , 
If the EventHandler is registered before the Informer data synchronization, then this PR can help determine whether the EventHandler is synchronized (processorListener.HasSync return true). But there is another scenario is to register the EventHandler after the Informer data synchronization is completed, and it is also expected to determine whether the EventHandler is synchronized. But this PR doesn't seem to support this scenario. Will this scenario be supported?

https://github.com/kubernetes/kubernetes/blob/fdd7b4ce5fde399d2a3878ca4f5d569e42dff017/staging/src/k8s.io/client-go/tools/cache/shared_informer.go#L577-L595

Reply:
Yeah this is one of the cases I want to fix/test.

Reply:
This is fixed in the latest push

Reply:
Thank you @lavalamp .
Not sure if the implementation of this place will have a race condition. For example, after the Cache data synchronization is completed, the EventHandler is registered. At this time, it traverses the data in the Cache, initiates an addNotification and calls syncTracker.Start, then syncTracker.count becomes to 1. When the addNotification is consumed, execute syncTracker.Finish, then syncTrakcer.count becomes to 0, just at this time an external goroutine called syncTracker.HasSynced, because upstream hasSynced, and syncTracker.count = 0, so syncTracker.HasSynced returns true.

Maybe can consider passing the total number of objects in the Cache directly to syncTracker.count before initiating addNotification.

WDYT?

Reply:
We hold the lock on the informer while adding these, which means no one can call the informer's HasSynced method untill we have added them all, which also means the handler's HasSynced method can't be called, since it calls the informer's HasSynced method. I can expand the comment here.

However now that you mention it I need to fix up the tracker to check things in the other order, I think.

Reply:
Actually if you read the comment it already mostly says this, but I will add your scenario in as another thing that's protected against this way.

Reply:
Yes... I overlooked that..😅